### PR TITLE
feat(addresses): implement addresses api endpoints

### LINKED ISSUE #9: BE | Auth | Address Management API
### 🧩 Feature:
User Address Management

### 🎯 Goal:
Build CRUD API endpoints for managing addresses, including support for setting a default shipping address and user address validation.

---

### ✅ Acceptance Criteria:
- [x] List, create, update, delete address endpoints
- [x] Set and update default shipping address
- [x] Ensure address is tied to correct user
- [x] Validate fields (postal code, city, etc.)

---

### 🧪 Testing:
- [ ] Unit tests for each CRUD operation
- [ ] Postman tests: edge cases, default toggle
- [ ] Unauthorized access test

### PR DESCRIPTION
**Pull request by mohamed952741**
## ✅ What Was Done

- [x] updated the model to allow only one default address
- [x] added Address serializer
- [x] implemented api views and urls
- [x] added tests to check default address handeling
- [x] updated requirements.txt to include pytest

## 🧪 How to Test

1. Run `docker compose up`
2. Run tests: `docker compose exec backend pytest`

### *DataBase needs migrations*
- [x] Database migration included
- [x] API affected

## 🔗 Related Issues

Closes #9 

## 📋 Checklist Before Merge

- [x] Code is self-documented and clean
- [x] Tests are added/updated
- [x] I rebased onto latest `Feature/auth/main` and resolved any conflicts

### CODE REVIEW

File: backend/addresses/models.py [hunk 1]
```diff
@@ -10,5 +9,10 @@ class Address(models.Model):
     country = models.CharField(max_length=100)
     is_default = models.BooleanField(default=False)
 
+    def save(self,*args,**kwargs):
+        if self.is_default:
+            Address.objects.filter(user=self.user, is_default=True).exclude(pk=self.pk).update(is_default=False)
+        super().save(*args, **kwargs)
+
```
Comment (AAEmara) [inline backend/addresses/models.py:16]: You really did a good job on customizing the save method, because it achieves the following:
  - [x] Checks if the new address to be saved is set to default address _(value is `True`)_, or not.
  - [x] If the address is set to default _(value is `False`)_, it unsets the other default address _(where value is `True`)_ for the same user.

File: backend/addresses/serializers.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        if is_default:
+            address.is_default = True
+            address.save(update_fields=["is_default"])
+
+            try:
+                profile = user.customerprofile
+                profile.default_shipping_address = address
+                profile.save(update_fields=["default_shipping_address"])
+            except CustomerProfile.DoesNotExist:
+                CustomerProfile.objects.create(user=user, default_shipping_address=address)
```
Comment (AAEmara) [inline backend/addresses/serializers.py]: Just a small update to meet the following business logic:
  - If the user has a `customer` role, they could add the address as a default address and update their
    `default_shipping_address` field.
  - The `create()` method should only be concerned with `creating an address`.
     - Creating a `CustomerProfile` is a user _onboarding concerns_, not an address-create concern.
    
Here my suggestion for the change:
```suggestion
            if user.role == "customer":
              try:
                  profile = user.customerprofile
                  profile.default_shipping_address = address
                  profile.save(update_fields=["default_shipping_address"])
              except CustomerProfile.DoesNotExist:
                  raise ValidationError("Customer profile must be created before assigning a default address.")
```

This ensures:
  - [x] That the address created as a default address only for users having a `customer` role.
  - [x] Doesn't blindly create a `CustomerProfile`, regardless of whether the `user.role` is `customer` or  not.
  - [x] Protects the business logic rules if validations are skipped for any reason.

File: backend/addresses/serializers.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    def validate_governorate(self, value):
+        if value not in EGYPT_GOVERNORATES:
+            raise serializers.ValidationError("Invalid governorate.")
+        return value
+
+    def validate_country(self, value):
+        if value.strip().lower() != "egypt":
+            raise serializers.ValidationError("Country must be 'Egypt'.")
+        return value
+
```
Comment (AAEmara) [inline backend/addresses/serializers.py:38]: Add a generic validation to make sure that only users with `customer` role, could create an address with `is_default=True`.

Here is my suggestion:
```suggestion

    def validate(self, attrs):
      user = self.context["request"].user
      is_default = attrs.get("is_default", False)
      
      if is_default and user.role != "customer":
        raise serializers.ValidationError("Only customers can set a default shipping address.")
      return attrs
```

This ensures:
  - No role other than `"customer"` could submit an address with `"is_default": true` inside the request's body.
  - Gives a 400 response before any DB queries are made.
  - Adds another layer of protection through validation.

File: backend/addresses/serializers.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        if is_being_defaulted:
+            updated_address.is_default = True
+            updated_address.save(update_fields=["is_default"])
+
+            try:
+                profile = updated_address.user.customerprofile
+                profile.default_shipping_address = updated_address
+                profile.save(update_fields=["default_shipping_address"])
+            except CustomerProfile.DoesNotExist:
+                CustomerProfile.objects.create(user=updated_address.user, default_shipping_address=updated_address)
```
Comment (AAEmara) [inline backend/addresses/serializers.py]: Again, make sure that we validate the user's role is a `"customer"` role.

Here is my suggestion.
```suggestion
            if updated_address.user.role == "customer":
              try:
                  profile = updated_address.user.customerprofile
                  profile.default_shipping_address = updated_address
                  profile.save(update_fields=["default_shipping_address"])
              except CustomerProfile.DoesNotExist:
                  raise ValidationError("Customer profile must be created before assigning a default address.")
```

File: backend/addresses/serializers.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        else:
+            try:
+                profile = updated_address.user.customerprofile
+                if profile.default_shipping_address == updated_address:
+                    profile.default_shipping_address = None
+                    profile.save(update_fields=["default_shipping_address"])
+            except CustomerProfile.DoesNotExist:
+                pass
```
Comment (AAEmara) [inline backend/addresses/serializers.py]: Again, we need to verify that the user's role is `"customer"`, where we could remove the default address if the
`is_default` value was `False`.

Here are my suggestions:
```suggestion
            if updated_address.user.role == "customer":
                try:
                    profile = updated_address.user.customerprofile
                    if profile.default_shipping_address == updated_address:
                        profile.default_shipping_address = None
                        profile.save(update_fields=["default_shipping_address"])
                except CustomerProfile.DoesNotExist:
                    raise ValidationError("Customer profile must exist to remove default address.")
```

File: backend/addresses/views.py [hunk 6]
```diff
@@ -1,3 +1,41 @@
...
+from rest_framework.decorators import action
+from rest_framework.response import Response
+from .models import Address
+from .serializers import AddressSerializer
+from accounts.models import CustomerProfile
 
-# Create your views here.
+class AddressViewSet(viewsets.ModelViewSet):
+    serializer_class = AddressSerializer
+    permission_classes = [permissions.IsAuthenticated]
```
Comment (AAEmara) [inline backend/addresses/views.py:10]: Great catch! We should validate the user's token and if they're authenticated or not before they can interact with the application's logic.

File: backend/addresses/views.py [hunk 7]
```diff
@@ -1,3 +1,41 @@
...
+
+    def get_queryset(self):
+        return Address.objects.filter(user=self.request.user)
+
+    def perform_create(self, serializer):
+        # The serializer handles setting the user and is_default logic
+        serializer.save()
```
Comment (AAEmara) [inline backend/addresses/views.py]: I believe we don't need to use `perform_create` in here, because we're not setting the user (not passing the `user` to the save method).

Also, the `perform_create` is called by default inside the View's `create` method.
```suggestion
```

This ensures:
  - [x] The view cleaner and avoids redundant method definitions.

File: backend/addresses/views.py [hunk 8]
```diff
@@ -1,3 +1,41 @@
...
+
+    @action(detail=True, methods=['post'], url_path='set-default')
```
Comment (AAEmara) [inline backend/addresses/views.py]: This `@action` decorator will need to be updated, after a slight change that affects different aspects of our code and business logic:
  - The API Endpoint will change from `POST /api/addresses/{id}/set-default/` into `PATCH /api/addresses/{id}/set-default/`
     - This ensures that we update the `default_shipping_address_id` specifically for a `"customer"`, and not any other field.
     - Makes a clear and distinct separation of concerns from other API endpoints.
     - Triggers the serializer's `update()` method instead of `create()` method, which makes sense in terms of business logic defined there.
     
Kindly, find the following suggestion:
```suggestion
    @action(detail=True, methods=['patch'], url_path='set-default')
```

This means that a request to the `` endpoint, would take something like this in its body:
```json
  {
    "is_default": true
  }
```

File: backend/addresses/views.py [hunk 9]
```diff
@@ -1,3 +1,41 @@
...
+
+        # Update customer profile
+        try:
+            profile = request.user.customerprofile
+            profile.default_shipping_address = address
+            profile.save(update_fields=["default_shipping_address"])
+        except CustomerProfile.DoesNotExist:
+            CustomerProfile.objects.create(user=request.user, default_shipping_address=address)
+
+        return Response({"detail": "Default shipping address set."})
```
Comment (AAEmara) [inline backend/addresses/views.py]: I could see a redunancy in the business logic defined inside the `AddressSerializer`'s `update()` method.

To keep the logic at the serializer side as defined previously and just use a thin layer of logic inside the `view`,
here is my suggestion:
```suggestion
    @action(detail=True, methods=['post'], url_path='set-default')
    def set_default(self, request, pk=None):
        address = self.get_object()

        if address.user != request.user:
            return Response({"detail": "Unauthorized"}, status=status.HTTP_403_FORBIDDEN)

         if request.user.role != "customer":
             return Response({"detail": "Only customers can set default addresses."}, status=403)

          # Call the serializer's update() with `is_default=True`, and eliminates redundancy.
          serializer = self.get_serializer(address, data={"is_default": True}, partial=True)
          serializer.is_valid(raise_exception=True)
          serializer.save()

        return Response({"detail": "Default shipping address set."})
```

This change, ensures:
  - [x] A separation of concerns, where the serializer includes only the main business logic
  - [x] A separation of concerns, where the model handles changes to addresse's is_default field.
  - [x] Avoids redundancy.

**Review by AAEmara — CHANGES_REQUESTED** [review]
Kindly review the comments and the suggested changes allover the `accounts app`.
Also, we could use a separate Database for testing, instead of creating those records resulted from reviewed tests inside the `developement database`.

Great effort @mohamed952741, keep moving forward!

**Comment by mohamed952741** [comment]
**Thanks for the review!**  
I’ve addressed the feedback as follows:  

- Updated validation to raise a `ValidationError` with clear, descriptive messages.  
- Adjusted logic so that `CustomerProfile.default_shipping_address` is set when a new default address is created.  
- Added tests to verify correct behavior for default address assignment and validation errors.  

Please let me know if anything else needs refining.

File: backend/addresses/views.py (hunk 8 above)
Comment (AAEmara) [reply backend/addresses/views.py]: Dear @mohamed952741, the `post` verb needs to be addressed.
Just replace the `post` verb with `patch`.

File: backend/addresses/serializers.py [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+                except CustomerProfile.DoesNotExist:
+                    raise serializers.ValidationError("Customer profile must be created before assigning a default address.")
+            else:
+                try:
+                    profile = updated_address.user.customerprofile
+                    if profile.default_shipping_address == updated_address:
+                        profile.default_shipping_address = None
+                        profile.save(update_fields=["default_shipping_address"])
+                except CustomerProfile.DoesNotExist:
+                    pass
```
Comment (AAEmara) [inline backend/addresses/serializers.py]: This section shall not fail silently; please raise a `ValidationError` when the `CustomerProfile.DoesNotExist` error isn't raised.
```suggestion
                except CustomerProfile.DoesNotExist:
                    raise ValidationError("Customer profile must exist to remove default address.")
```

File: backend/addresses/serializers.py (hunk 5 above)
Comment (AAEmara) [reply backend/addresses/serializers.py]: Kindly @mohamed952741, this section shall not fail silently, please find my suggestion in the new review.

**Review by AAEmara — CHANGES_REQUESTED** [review]
Just minor changes are left to be done, and we're good to go.

**Comment by mohamed952741** [comment]
**Thanks for the follow up review!**  
I’ve addressed the feedback as follows:  

- Added ValidationError("Customer profile must exist to update address.") to line 87 in serializers.py instead of passing silently.
- Updated the post method of the action to patch  

Please let me know if anything else needs refining.

File: backend/addresses/serializers.py [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+                except CustomerProfile.DoesNotExist:
+                    raise serializers.ValidationError("Customer profile must be created before assigning a new default address.")
+            else:
+                try:
+                    profile = updated_address.user.customerprofile
+                    if profile.default_shipping_address == updated_address:
+                        profile.default_shipping_address = None
+                        profile.save(update_fields=["default_shipping_address"])
+                except CustomerProfile.DoesNotExist:
+                    raise serializers.ValidationError("Customer profile must exist to update address.")
```
Comment (AAEmara) [inline backend/addresses/serializers.py:87]: @mohamed952741 Great job.

File: backend/addresses/views.py [hunk 12]
```diff
@@ -1,3 +1,30 @@
...
 
-# Create your views here.
+class AddressViewSet(viewsets.ModelViewSet):
+    serializer_class = AddressSerializer
+    permission_classes = [permissions.IsAuthenticated]
+
+    def get_queryset(self):
+        return Address.objects.filter(user=self.request.user)
+
+    @action(detail=True, methods=['patch'], url_path='set-default')
```
Comment (AAEmara) [inline backend/addresses/views.py:15]: Great job @mohamed952741.

**Review by AAEmara — APPROVED** [review]
You did a great deal of worked, and it have finally paid of with approval.
Keep doing amazing work @mohamed952741.
